### DISCUSSION TITLE
Role based authorization

### DISCUSSION THREAD

**Discussion by ghost** [ISIL-ESTE/Student-Workflow-Organizer#66] [Q&A] [answered]
Why we are using 2 restrictTo functions in admin controller?

```js 
router.use(authController.restrictTo('ADMIN', 'SUPER_ADMIN'));

router
  .route('/user/:id')
  .get(userController.getUser)
  .patch(userController.updateUser)
  .delete(userController.deleteUser);
router.get('/users', userController.getAllUsers);


router.put(
  '/authorize-or-restrict/:userId',
  restrictTo(Actions.UPDATE_USER),
  authorizeOrRestrict
);
```

@bellaabdelouahab

**Comment by bellaabdelouahab** [comment]
Let me see,
I think we added this to prevent admins from restricting each other,
And also I think there is a diff between restricting to a role and restricting to and authority, maybe we need to chamge the names to know which one which
@muttaqin1

**Reply by ghost** [reply]
The restrictTo func from the authController  blocks user with custom roles to access in this routes.
We can point a user if he is admin or not by their authorities. There is nothing called roles in this approach. Roles are just a set of authorities and restrictions.
It will be better to remove that middleware.
And we can farther discuss how to prevent one admin to restrict another admin.
 
What do you think?
@bellaabdelouahab

**Reply by bellaabdelouahab** [reply]
I don't know if this is the best approach to handle roles, but I'd love to hear more about your idea.
@muttaqin1

**Reply by ghost** [reply]
I believe we dont have any other way.
I suggest you to create a custom role 'Moderator' and provide all the authorities of the admin. Then try to access that route.You cant access because that middleware only gives access to admin and superAdmin.
@bellaabdelouahab

**Reply by bellaabdelouahab** [reply]
I understand you but we still need to validate the roles other ways we will have to verify every authority,
Search one open source if you found that this is common practice, you can go ahead and implement it.
@muttaqin1

**Reply by ghost** [reply, marked as answer]
Yes, Thats correct we have to validate the roles or every authority.
The logic to validate every authority is already implemented.

so we can do this:
We will validate **User** role by the **Role** and all the upper level roles like **Admin**, **superAdmin** will be validated by **authorities**.
Now just we have to remove the authController restrictTo middleware.
By doing this we can achieve a great admin management.
@bellaabdelouahab

**Reply by bellaabdelouahab** [reply]
@muttaqin1 great 👌

**Comment by ghost** [comment]
Can you do the task? This is a small task, It wont take more than five minutes. Just remove the function call.
@bellaabdelouahab

**Comment by bellaabdelouahab** [comment]
sure 👌
